drivers/lis2dh12/lis2dh12.c

```diff
@@ -264,14 +274,283 @@ int lis2dh12_read_int_src(const lis2dh12_t *dev, uint8_t *data, uint8_t int_line
 
     return LIS2DH12_OK;
 }
+
+int lis2dh12_configure_fifo(const lis2dh12_t *dev, const lis2dh12_fifo_t *config) {
+
+    assert(dev && config);
+
+    LIS2DH12_CTRL_REG5_t reg5 = {0};
+    LIS2DH12_FIFO_CTRL_REG_t fifo_reg = {0};
+
+    if (config->FIFO_mode != LIS2DH12_FIFO_MODE_BYPASS) {
+        reg5.bit.FIFO_EN = 1;
+    }
+    fifo_reg.bit.TR = config->FIFO_set_INT2;
+    fifo_reg.bit.FM = config->FIFO_mode;
+    fifo_reg.bit.FTH = config->FIFO_watermark;
+
+    _acquire(dev);
+    _write(dev, REG_CTRL_REG5, reg5.reg);
+    _write(dev, REG_FIFO_CTRL_REG, fifo_reg.reg);
+    _release(dev);
+
+    return LIS2DH12_OK;
+}
+
+int lis2dh12_restart_fifo(const lis2dh12_t *dev) {
+
+    assert(dev);
+
+    _acquire(dev);
+    uint8_t reg5 = _read(dev, REG_CTRL_REG5);
+    LIS2DH12_FIFO_CTRL_REG_t fifo_reg = {0};
+    fifo_reg.reg = _read(dev, REG_FIFO_CTRL_REG);
+
+    uint8_t fifo_mode_old = fifo_reg.bit.FM;
+    fifo_reg.bit.FM = LIS2DH12_FIFO_MODE_BYPASS;
+
+    /* switch to Bypass mode */
+    _write(dev, REG_FIFO_CTRL_REG, fifo_reg.reg);
+
+    fifo_reg.bit.FM = fifo_mode_old;
+
+    _write(dev, REG_CTRL_REG5, reg5);
+    _write(dev, REG_FIFO_CTRL_REG, fifo_reg.reg);
+    _release(dev);
+
+    return LIS2DH12_OK;
+}
+
+int lis2dh12_read_fifo_src(const lis2dh12_t *dev, uint8_t *data) {
+
+    assert(dev && data);
+
+    _acquire(dev);
+    *data = _read(dev, REG_FIFO_SRC_REG);
+    _release(dev);
+
+    return LIS2DH12_OK;
+}
+
+uint8_t lis2dh12_read_fifo_data(const lis2dh12_t *dev, lis2dh12_fifo_data_t *fifo_data, uint8_t number) {
+
+    assert(dev && fifo_data);
+    /* check max FIFO length */
+    assert(number <= 32);
+
+    _acquire(dev);
+    /* check if number is available */
+    LIS2DH12_FIFO_SRC_REG_t src_reg = {0};
+    src_reg.reg = _read(dev, REG_FIFO_SRC_REG);
+
+    if (src_reg.bit.FSS <= number) {
+        number = src_reg.bit.FSS;
+    }
+
+    if (src_reg.bit.EMPTY) {
+        return 0;
+    }
+
+    uint8_t regs[6] = {REG_OUT_X_L, REG_OUT_X_H, REG_OUT_Y_L, REG_OUT_Y_H, REG_OUT_Z_L, REG_OUT_Z_H};
+    uint8_t fifo_data_raw[6*number];
+
+    for (uint8_t i = 0; i < number*6; i++){
+        fifo_data_raw[i] = _read(dev, regs[i % 6]);
```

Comment:
```suggestion
        fifo_data[i].FIFO_data_X = _read(dev, REG_OUT_X_L)
                                 | _read(dev, REG_OUT_X_H) << 8;
        fifo_data[i].FIFO_data_Y = _read(dev, REG_OUT_Y_L) 
                                 | _read(dev, REG_OUT_Y_H) << 8;
        fifo_data[i].FIFO_data_Z = _read(dev, REG_OUT_Z_L) 
                                 | _read(dev, REG_OUT_Z_H) << 8;
```

Reply:
You can also do the conversion here directly, no need to iterate the array a second time.

Reply:
this does not work, because the data is stored in twos complement

Reply:
Huh? Which CPU *doesn't* use two's complement?

Reply:
maybe thats a good why to do. Because the resolution does change by the power mode. So to check which power mode is set, we need to read out the REG and then change the bits for sign extend.
I think the old solution does support this neither. (need to change this in `lis2dh12_read()` as well)

Reply:
So want to add this? :)

Reply:
Think this is done, right? ^^